core/src/main/java/org/infinispan/context/Flag.java

```diff
@@ -284,6 +295,44 @@
       }
    }
 
+   public static Set<Flag> addFlag(Set<Flag> flags, Flag newFlag) {
+      if (flags == null) {
+         return EnumSet.of(newFlag);
+      }
+      if (flags.contains(newFlag)) {
+         return flags;
+      }
+      EnumSet<Flag> copy = EnumSet.copyOf(flags);
+      copy.add(newFlag);
+      return copy;
+   }
+
+   public static Set<Flag> addFlags(Set<Flag> flags, Flag... newFlags) {
+      if (newFlags == null || newFlags.length == 0) {
+         return flags;
+      }
+      EnumSet<Flag> copy;
+      if (flags == null) {
+         copy = EnumSet.noneOf(Flag.class);
+      } else {
+         copy = EnumSet.copyOf(flags);
+      }
+      Collections.addAll(copy, newFlags);
+      return copy;
+   }
+
+   public static Set<Flag> addFlags(Set<Flag> flags, Set<Flag> newFlags) {
+      if (newFlags == null || newFlags.isEmpty()) {
+         return flags;
+      }
+      if (flags == null) {
```

Comment:
`copyOf` isn't really more expensive than `noneOf`, so I'd rather avoid the `isEmpty` check.